src/main/asciidoc/docs/testng_listeners.adoc

```diff
@@ -214,3 +214,142 @@ Finishing
 
 This mechanism allows you to apply the same set of listeners to an entire organization just by adding a jar file to the classpath, instead of asking every single developer to remember to specify these listeners in their `testng.xml` file.
 
+==== Ordering listeners in TestNG
+
+TestNG now allows you to control the order in which listeners are executed.
+
+This is particularly useful when you have multiple listeners that are altering the test result states and so you would like to ensure that they do so in some deterministic order.
+
+Following is how we can get this done.
+
+* To start with, you would need to build an implementation of the interface `org.testng.ListenerComparator`
+* Now this implementation needs to be plugged into TestNG via the configuration parameter `-listenercomparator`
+
+CAUTION: TestNG orders only user's listeners and which are not part of the exclusion list that can be specified via the JVM argument `-Dtestng.preferential.listeners.package` (Multiple fully qualified class names can be specified as comma separated values). If nothing is specified, TestNG by default excludes all the IntelliJ IDEA listeners under the package `com.intellij.rt.*`.
+
+Let's see an example.
+
+* Let's create a custom annotation that will help us define the order in which *our* listeners should be invoked.
+
+[source, java]
+
+----
+@Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+@Target({ TYPE})
+
+public @interface RunOrder {
+  int value();
+}
+----
+
+* Now let's create an implementation of `org.testng.ListenerComparator` which honours this custom annotation that we just now created.
+
+[source, java]
+
+----
+import org.testng.ITestNGListener;
+import org.testng.ListenerComparator;
+
+public class AnnotationBackedListenerComparator implements ListenerComparator {
+
+  @Override
+  public int compare(ITestNGListener l1, ITestNGListener l2) {
+    RunOrder one = l1.getClass().getAnnotation(RunOrder.class);
+    RunOrder two = l2.getClass().getAnnotation(RunOrder.class);
+    if (one != null && two != null) {
+      return Integer.compare(one.value(), two.value());
+    }
+    //Add all other listeners which don't have the @RunOrder annotation to the end so that they get executed
```

Comment:
Sure. But that would make the test code a bit more verbose because I would have to check both the listener objects to see if they had null annotation and then define a default value. Since this was just for illustration purposes, I thought this would suffice.

Reply:
I understand but it is not a logic we can recommend. Create an utils method that extract value from the object and you'll avoid the verbosity.

Reply:
Have fixed this